### PR TITLE
Generate most of the gRPC wrappers in client

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
Generate most of the code in `client/...` instead of having to copy/paste when adding new rpcs.

Some things that looked wrong were fixed. There are a few `TODO`s in the generator that note irregularities, e.g. some history client methods that didn't use `executeWithRedirect`.

Unfortunately the reflect package uses lexicographic order instead of declaration order, so it's really hard to compare the before and after. I might try to write a script that splits all functions into separate files and generate a diff that way. Open to other ideas.

**Why?**
Boilerplate is bad. Codegen is... not good, but at least it's the Go way.

**How did you test it?**
integration tests

**Potential risks**
If there's a substantial difference in implementation (not just metrics), this could break things in subtle ways.

### CODE REVIEW

File: cmd/tools/rpcwrappers/main.go [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+		// these do forwarding stats. too complicated.
+		"metricsClient.matching.AddActivityTask":       true,
+		"metricsClient.matching.AddWorkflowTask":       true,
+		"metricsClient.matching.PollActivityTaskQueue": true,
+		"metricsClient.matching.PollWorkflowTaskQueue": true,
+		"metricsClient.matching.QueryWorkflow":         true,
+	}
+)
+
+func writeCopyright(w io.Writer) {
```
Comment (yux0) [inline cmd/tools/rpcwrappers/main.go:114]: We have a licensegen.go. is it possible to reuse it?

Comment (dnr) [reply cmd/tools/rpcwrappers/main.go]: sure, I made it read the root LICENSE file similar to that

**Comment by dnr** [comment]
I wrote a little script to split out things by function so it's easier to see the actual diff. This commit corresponds to the difference between master and this PR: https://github.com/dnr/temporal/commit/7e65d811531dc4bce7e383e0935f641c0d3b3b9b

**Comment by Sushisource** [comment]
Wahoo! Thanks for this David

**Review by yiminc — APPROVED** [review]

File: client/admin/client.go [hunk 2]
```diff
@@ -22,14 +22,16 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+// Generates all three generated files in this package:
+//go:generate go run ../../cmd/tools/rpcwrappers -service admin
```
Comment (alexshtin) [inline client/admin/client.go:26]: This is first time we use `go:generate` for something different than mocks. `make mocks` will execute this also, which is unexpected.

I think this should be leveraged:
```
Go generate accepts one specific flag:

        -run=""
                if non-empty, specifies a regular expression to select
                directives whose full original source text (excluding
                any trailing spaces and final newline) matches the
                expression.
```

Comment (alexshtin) [reply client/admin/client.go:26]: And dedicated `Makefile` target is needed for this. And it needs to be included in `ci-build`.

File: cmd/tools/rpcwrappers/main.go [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+}
+`)
+	// TODO: some history methods did not touch metrics. should we preserve this?
+	// DescribeHistoryHost
+	// RemoveTask
+	// CloseShard
+	// GetShard
+	// RebuildMutableState
+	// DescribeMutableState
+	// TODO: DeleteWorkflowExecution didn't work like the others in history (the code looked
```
Comment (alexshtin) [inline cmd/tools/rpcwrappers/main.go:398]: `DeleteWorkflowExecution` should work the same as other methods i.e. `TerminateWorkflowExecution`.

File: cmd/tools/rpcwrappers/main.go [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+	return c.client.{{.Method}}(ctx, request, opts...)
+}
+`)
+	// TODO: some history methods did not touch metrics. should we preserve this?
+	// DescribeHistoryHost
+	// RemoveTask
+	// CloseShard
+	// GetShard
+	// RebuildMutableState
+	// DescribeMutableState
```
Comment (alexshtin) [inline cmd/tools/rpcwrappers/main.go:397]: I would vote for unification. Let's create metrics, no harm from it.

File: cmd/tools/rpcwrappers/main.go [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+	if i < 0 {
+		panic("unknown service")
+	}
+	svc := services[i]
+
+	licenseText := readLicenseFile(*licenseFlag)
+
+	callWithFile(svc.clientGenerator, svc, "client", licenseText)
+	callWithFile(svc.metricGenerator, svc, "metricClient", licenseText)
+	callWithFile(svc.retryableGenerator, svc, "retryableClient", licenseText)
```
Comment (alexshtin) [inline cmd/tools/rpcwrappers/main.go:474]: Let's follow go best practices for all new file names:
```suggestion
	callWithFile(svc.metricGenerator, svc, "metric_client", licenseText)
	callWithFile(svc.retryableGenerator, svc, "retryable_client", licenseText)
```

File: develop/docs/new-rpcs.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+Adding new RPCs to any of the server roles is currently a little involved.
+Here's a list of what you'll need to modify:
+
+1. Add RPC definitions to proto files. Either in the `api` repo (for frontend)
+   or here in `proto/internal` (for history/matching).
+2. Update `go.temporal.io/api` to pick up the new generated files (for frontend)
+   or generate them here with `make proto` (for history/matching).
+3. Add metric scope defs for client-side metrics to `common/metrics/defs.go`
+   (this is going away soon).
+4. `make go-generate` to generate wrappers in `client`.
```
Comment (alexshtin) [inline develop/docs/new-rpcs.md:12]: so it should be `make clients`

**Review by alexshtin — CHANGES_REQUESTED** [review]

File: cmd/tools/rpcwrappers/main.go [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+}
+
+func pathToField(t reflect.Type, name string, path string, maxDepth int) string {
+	if t.Kind() != reflect.Struct || maxDepth <= 0 {
+		return ""
+	}
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.Name == name {
+			return path + "." + name
```
Comment (dnr) [inline cmd/tools/rpcwrappers/main.go:129]: I'm going to change this to use Get() for safety

File: client/admin/client.go (hunk 2 above)
Comment (alexshtin) [reply client/admin/client.go:26]: OK, then don't forget to remove `mocks` target from `Makefile` in that PR too.

File: cmd/tools/rpcwrappers/main.go [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+	}
+	client, err := c.getClientForWorkflowID(request.NamespaceId, taskToken.GetWorkflowId())
+`, path)
+	}
+	// slice needs a tiny bit of extra handling for namespace
+	if path := pathToField(t, "TaskInfos", "request", 1); path != "" {
+		return fmt.Sprintf(`// All workflow IDs are in the same shard per request
+	client, err := c.getClientForWorkflowID(%s[0].NamespaceId, %s[0].WorkflowId)`, path, path)
+	}
+	panic("I don't know how to get a client from a " + t.String())
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply cmd/tools/rpcwrappers/main.go:166]: It is up to you. It is your baby after all.

File: cmd/tools/rpcwrappers/main.go (hunk 5 above)
Comment (alexshtin) [reply cmd/tools/rpcwrappers/main.go]: yes, please rename other too.

Actually I believe we need to rename all our files to the snake_case.

**Review by alexshtin — APPROVED** [review]

**Pull request merged**